src/pytorch_lightning/utilities/data.py

```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
     batch_sampler = getattr(dataloader, "batch_sampler")
     is_predicting = mode == RunningStage.PREDICTING
     # checking the batch sampler type is different than PyTorch default.
-    if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
-        batch_sampler = type(batch_sampler)(
-            sampler,
-            batch_size=batch_sampler.batch_size,
-            drop_last=(False if is_predicting else batch_sampler.drop_last),
-        )
+    batch_cls = type(batch_sampler)
+    if batch_sampler is not None and (batch_cls is not BatchSampler or is_predicting):
+        if hasattr(batch_sampler, "__pl_saved_args"):
+            args = list(batch_sampler.__pl_saved_args)
+            kwargs = batch_sampler.__pl_saved_kwargs
+            arg_names = batch_sampler.__pl_saved_arg_names
+
+            if is_predicting:
+                success, args, kwargs = _inject_arg_to_saved(args, kwargs, arg_names, "drop_last", False)
+                if not success:
+                    rank_zero_warn(
+                        "Trying to inject `drop_last=False` into batch sampler since you are predicting, however it "
+                        f"seems the class `{batch_cls}` does not support it. Your predictions might be incomplete. "
```

Comment:
The class is a class so this needs to use the qualname to avoid printing something like

`<class '__main__.Foo'>"`

```suggestion
                        f"seems the class `{batch_cls.__qualname__}` does not support it. Your predictions might be incomplete. "
```

Same comment applies to other usages of the class in a string